### ISSUE TITLE
Allow emscripten_fetch in Asyncify

### ISSUE THREAD

**Issue by RReverser** [emscripten-core/emscripten#12255]
Currently there is a perceived inconsistency between APIs like `emscripten_sleep` and `emscripten_fetch`: while both perform sort of an async I/O, one of them works with Asyncify, while another works only by compiling code in a separate thread.

Technically, there is no reason for this discrepancy, and it should be possible to allow APIs like `emscripten_fetch` with just Asyncify if user wants it to run on the main thread instead of a separate one.

Perhaps this is a more general issue though and there should be some shared mechanism for all such APIs to choose between Asyncify and thread-based blocking depending on compile flags?

**Comment by kripken** [comment]
I agree it would be better to support all APIs with Asyncify. It does take a little manual work for each which is why that's not the case yet.

Also no one is actively working on Fetch atm. @juj has plans to create something lower-level and better. We can add Asyncify support more easily for that API, I expect.

**Comment by RReverser** [comment]
I'm a bit worried about this getting into https://xkcd.com/927/ territory, where we simply will end up with 3 APIs that work in different modes :)

We already have `emscripten_wget` and `emscripten_fetch`, where both do sorts of network requests, but cover different use-cases, and one works only with Asyncify and another only with proxy-to-thread.

I think ideally we should still consolidate existing APIs to work under either of those modes, regardless of future additions.

**Comment by RReverser** [comment]
Bump.

**Comment by RReverser** [comment]
Just need to be careful with various conditions: off the main thread emscripten_thread_sleep is more efficient as it just uses atomic.wait; on the main thread emscripten_sleep is better than busy-loop, but only works when user opted in to Asyncify + when used implicitly, can have some surprising effects due to exports becoming async too now.

**Comment by RReverser** [comment]
Probably expressed myself in a bit confusing manner, so just to clarify: yes, I certainly agree it would be great for emscripten_thread_sleep / usleep / etc. to use Asyncify, just need to make sure we use it only 1) on the main thread and 2) if the user explicitly opted in to Asyncify.

**Comment by sbc100** [comment]
Yup, thanks for extra details there.. certainly seems more complex than I originally though.

**Comment by RReverser** [comment]
Heh yeah which brings me back to

> Perhaps this is a more general issue though and there should be some shared mechanism for all such APIs to choose between Asyncify and thread-based blocking depending on compile flags?

It would be cool to solve this problem once for all APIs to avoid repeating complexity. So that there would be a single internal API which you could ask to do something asynchronous, and it would choose a mechanism between Asyncify / pthread / block on the same thread based on user-provided flags.

sleep and fetch seem like excellent examples to cover first for a PoC.
